docs/developers/developer-reference/smart-contract-developer-best-practices.md

```diff
@@ -247,28 +248,82 @@ For the same token ID and skipped nonce, the encodings look like this:
+The first approach looks a lot simpler, just a `ManagedVec` of tuples. But, the implications are quite devastating, both for performance and usability. To use the first endpoint, with the pairs (TOKEN-123456, 5, 100) and (TOKEN-123456, 10, 500), the call data would have to look like this:
 `myVarArgsEndpoint@0000000c_544f4b454e2d313233343536_0000000000000005_00000001_64_0000000c_544f4b454e2d313233343536_000000000000000a_00000002_01f4`
 
 Note: Above, we've separated the parts with `_` for readability purposes only. On the real blockchain, there would be no underscores, everything would be concatenated.
 
 As you can see, that endpoint is very hard to work with. All arguments have to be passed into this big chunk, with nested encoding, which also adds additional lengths for the `TokenIdentifier` (i.e. the 0000000c in front, which is length 12) and for `BigUint` (i.e. the length in bytes).  
 
-For the `VarArgs` approach, this endpoint is a lot easier to use. For the same arguments, the call data looks like this:
+For the `ManagedVarArgs` approach, this endpoint is a lot easier to use. For the same arguments, the call data looks like this:
 `myVarArgsEndpoint@544f4b454e2d313233343536@05@64@544f4b454e2d313233343536@0a@01f4`
 
 The call data is a lot shorter, and it's much more readable, and as we use top-encoding instead of nested-encoding, there's no need for lengths either.  
 
+## The dynamic allocation problem
+
+The main difference between the base Rust types (like `Vec<T>`) and their managed counterparts, provided by the Rust framework (like `ManagedVec<T>`) lies in their memory allocation. Base Rust types use dynamic allocation on the heap, which in simple terms, means the program (in this case, the smart contract), keeps asking for more and more memory from the operating system. For smalls collections, this doesn't matter much, but for bigger collection, this can make it so the contract call fails.  
+
+The main issue is that the Rust runtime is quite eager with its dynamic memory allocation, and asks for more memory than it actually needs. For normal programs, this is great for performance, but for smart contracts, where every instruction costs gas, can be quite impactful, on both cost and even runtime failures.  
+
+The alternative is to use managed types. All managed types, like `BigUint`, `TokenIdentifier`, `ManagedBuffer` etc. store all their memory inside the Elrond Virtual Machine, which is the program that executes smart contracts and provides the API functions. The managed types only store a `handle`, which is a `u32` index. So whenever you have to add two `BigUint`s for example, you only pass the three handles: the result, the first operand, and the second operand. This way, there is very little data being passed around, which in turn makes everything cheaper. And since these types only store a handle, their memory allocation is fixed size, so it can be allocated on the stack instead of having to be allocated on the heap.  
```

Comment:
"is to use **managed types**" (bold formatting).

Reply:
A bit more specificity (optionally).

_The managed types only store a `handle`_ + _within the contract memory, while the actual payload resides in VM's memory_.

_allocated on the stack_ + _of the executing contract_.

Reply:
Fixed.